File: src/main/java/com/developerphil/adbidea/ui/ModuleChooserDialogHelper.java
package com.developerphil.adbidea.ui;

import com.google.common.collect.Lists;
import com.intellij.ide.util.PropertiesComponent;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.module.ModuleType;
import com.intellij.openapi.project.Project;
import java.util.Collections;
import java.util.List;
import org.jetbrains.android.facet.AndroidFacet;

public final class ModuleChooserDialogHelper {

    public static final String SELECTED_MODULE_PROPERTY = ModuleChooserDialogHelper.class.getCanonicalName() + "-SELECTED_MODULE";
    public static final String DEFAULT_MODULE_PROPERTY = ModuleChooserDialogHelper.class.getCanonicalName() + "-DEFAULT_MODULE";

    public static final String DO_NOT_SELECT_THE_DEFAULT_MODULE = "Do not select the default module";


    private ModuleChooserDialogHelper() {
    }

    public static AndroidFacet showDialogForFacets(Project project, List<AndroidFacet> facets,boolean isSetDefault) {
        List<Module> modules = Lists.newArrayList();
        List<String> modulesName = Lists.newArrayList();
        String previousModuleName = getPreviousModuleName(project);
        String defaultModuleName = getDefaultModuleName(project);
        List<String> previousSelectedModule = null;
        for (AndroidFacet facet : facets) {
            Module module = facet.getModule();
            String name = module.getName();
            if (!isSetDefault && name.equals(defaultModuleName)) {
                return facet;
            }
            modules.add(module);
            modulesName.add(name);
            if (name.equals(previousModuleName)) {
                previousSelectedModule = Lists.newArrayList(name);
            } else if (isSetDefault && name.equals(defaultModuleName)) {
                previousSelectedModule = Lists.newArrayList(name);
            }
        }

        if (isSetDefault && previousSelectedModule == null) {
            previousSelectedModule = Lists.newArrayList(DO_NOT_SELECT_THE_DEFAULT_MODULE);
        }

        Collections.sort(modules, (o1, o2) -> o1.getName().compareToIgnoreCase(o2.getName()));
        Collections.sort(modulesName, String :: compareToIgnoreCase);
        if (isSetDefault) {
            modulesName.add(DO_NOT_SELECT_THE_DEFAULT_MODULE);
        }
        String titleSelectDefault = "Choose Default Module";
        if (!Utils.isEmpty(defaultModuleName)) {
            titleSelectDefault += ",Current module is :" + defaultModuleName;
        }
        MyChooseModulesDialog dialog = new MyChooseModulesDialog(project, modulesName, isSetDefault? titleSelectDefault :"Choose Module", isSetDefault?"Set the default module for each operation":"", ModuleType.get(modules.get(0)).getIcon());
        dialog.setSingleSelectionMode();
        if (previousSelectedModule != null) {
            dialog.selectElements(previousSelectedModule);
        }
        dialog.show();

        List<String> chosenElements = dialog.getChosenElements();
        if (chosenElements.isEmpty()) {
            return null;
        }

        String chosenModule = chosenElements.get(0);
        if (isSetDefault) {
            if (chosenModule.equals(DO_NOT_SELECT_THE_DEFAULT_MODULE)) {
                saveDefaultModuleName(project,"");
                return null;
            } else {
                saveDefaultModuleName(project,chosenModule);
            }
        } else {
            saveModuleName(project, chosenModule);
        }
        int chosenModuleIndex = modulesName.indexOf(chosenModule);
        return facets.get(chosenModuleIndex);
    }

    private static void saveDefaultModuleName(Project project, String moduleName) {
        final PropertiesComponent properties = PropertiesComponent.getInstance(project);
        properties.setValue(DEFAULT_MODULE_PROPERTY, moduleName);
    }

    private static void saveModuleName(Project project, String moduleName) {
        final PropertiesComponent properties = PropertiesComponent.getInstance(project);
        properties.setValue(SELECTED_MODULE_PROPERTY, moduleName);
    }

    private static String getPreviousModuleName(Project project) {
        final PropertiesComponent properties = PropertiesComponent.getInstance(project);
        return properties.getValue(SELECTED_MODULE_PROPERTY);
    }

    private static String getDefaultModuleName(Project project) {
        final PropertiesComponent properties = PropertiesComponent.getInstance(project);
        return properties.getValue(DEFAULT_MODULE_PROPERTY);
    }


}

File: src/main/kotlin/com/developerphil/adbidea/ui/ModuleChooserDialogHelper.kt
package com.developerphil.adbidea.ui

import com.intellij.ide.util.PropertiesComponent
import com.intellij.openapi.module.Module
import com.intellij.openapi.project.Project
import com.intellij.openapi.roots.ui.configuration.ChooseModulesDialog
import com.intellij.ui.table.JBTable
import com.intellij.util.ui.UIUtil
import org.jetbrains.android.facet.AndroidFacet
import java.awt.Component
import java.awt.Dimension
import java.awt.geom.Dimension2D
import javax.swing.JTable

object ModuleChooserDialogHelper {

    fun showDialogForFacets(project: Project, facets: List<AndroidFacet>): AndroidFacet? {
        val modules = facets.map { it.module }
        val previousModuleName = getSavedModuleName(project)
        val previousSelectedModule = modules.firstOrNull { it.name == previousModuleName }

        val selectedModule = showDialog(project, modules, previousSelectedModule) ?: return null
        saveModuleName(project, selectedModule.name)
        return facets[modules.indexOf(selectedModule)]
    }

    private fun showDialog(project: Project, modules: List<Module>, previousSelectedModule: Module?): Module? {
        with(ChooseModulesDialog(project, modules, "Choose Module", "")) {
            setSingleSelectionMode()
            getSizeForTableContainer(preferredFocusedComponent)?.let {
                // Set the height to 0 to allow the dialog to resize itself to fit the content.
                setSize(it.width, 0)
            }
            previousSelectedModule?.let { selectElements(listOf(it)) }
            return showAndGetResult().firstOrNull()
        }
    }

    // Fix an issue where the modules dialog is not wide enough to display the whole module name.
    // This code is lifted from com.intellij.openapi.ui.impl.DialogWrapperPeerImpl.MyDialog.getSizeForTableContainer
    private fun getSizeForTableContainer(component: Component?): Dimension? {
        if (component == null) return null
        val tables = UIUtil.uiTraverser(component).filter(JTable::class.java)
        if (!tables.isNotEmpty) return null
        val size = component.preferredSize
        for (table in tables) {
            val tableSize = table.preferredSize
            size.width = size.width.coerceAtLeast(tableSize.width)
        }
        size.width = size.width.coerceIn(600, 1000)
        return size
    }

    private fun saveModuleName(project: Project, moduleName: String) {
        PropertiesComponent.getInstance(project).setValue(SELECTED_MODULE_PROPERTY, moduleName)
    }

    private fun getSavedModuleName(project: Project): String? {
        return PropertiesComponent.getInstance(project).getValue(SELECTED_MODULE_PROPERTY)
    }
}

private val SELECTED_MODULE_PROPERTY = ModuleChooserDialogHelper::class.java.canonicalName + "-SELECTED_MODULE"